Supporting automated sip generation for qgsdataitem and qgsvector classes

doublebyte1 opened the pull request:
- Removed qgsdataitem.sip and qgsvector.sip from auto_sip.blacklist
 - Annotated qgsdataitem.h for sip
 - regenerated automatically qgsdataitem.sip and qgsvector.sip

## Description
This PR supports the automated generation of the sip files for the QgsVector and QsDataItem classes. In the case of the QgsVector it was not necessary to change anything, apart from removing it from the auto_sip.blacklist. For the case of the QgsVector class, it was necessary to annotate the header with the relevant sip notations (e.g.: SIP_TRANSFERTHIS, SIP_FACTORY, etc).

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

doublebyte1 committed:
- Supporting automated sip generation for qgsdataitem and qgsvector classes
 - Removed qgsdataitem.sip and qgsvector.sip from auto_sip.blacklist
 - Annotated qgsdataitem.h for sip
 - regenerated automatically qgsdataitem.sip and qgsvector.sip

doublebyte1 committed:
- Supporting automated sip generation for qgsdataitem and qgsvector classes
 - Removed qgsdataitem.sip and qgsvector.sip from auto_sip.blacklist
 - Annotated qgsdataitem.h for sip
 - regenerated automatically qgsdataitem.sip and qgsvector.sip

doublebyte1 commented:
@3nids @m-kuhn Please review the qgsdataitem.h, in case I missed anything :-) Thanks!

m-kuhn commented on python/core/qgsvector.sip:
```diff
@@ -1,47 +1,163 @@
+%End
 
-    QgsVector& operator+=( QgsVector other );
+    QgsVector &operator+=( QgsVector other );
```
Not strictly required, but can you change the method to

    QgsVector &operator+=( const QgsVector &other );

doublebyte1 replied:
done

m-kuhn commented on python/core/qgsvector.sip:
```diff
@@ -1,47 +1,163 @@
+ :rtype: QgsVector
+%End
 
-    QgsVector& operator-=( QgsVector other );
```
Not strictly required, but can you change the method to

QgsVector &operator-=( const QgsVector &other );

doublebyte1 replied:
done

nyalldawson replied:
Actually, this was intentional. QgsVector is small and trivially copiable, so it's faster to pass by value then reference here (add elsewhere in the class)

m-kuhn replied:
can we merge it with the `const QgsVector other` state (i.e. without the reference `&` part in there)?

nyalldawson replied:
Sure

m-kuhn commented on python/core/qgsvector.sip:
```diff
@@ -1,47 +1,163 @@
+ :rtype: QgsVector
+%End
 
     QgsVector operator-( QgsVector other ) const;
```
Not strictly required, but can you change the method to

QgsVector operator-( const QgsVector &other );

doublebyte1 replied:
done

m-kuhn commented on python/core/qgsdataitem.sip:
```diff
@@ -1,28 +1,46 @@
+
+
+
+typedef QgsDataItem *dataItem_t( QString, QgsDataItem * );
```
This isn't needed in python:

    typedef QgsDataItem *dataItem_t( QString, QgsDataItem * ) SIP_SKIP;

doublebyte1 replied:
done

m-kuhn commented on python/core/qgsdataitem.sip:
```diff
@@ -1,28 +1,46 @@
+    if ( sipCpp->inherits( "QgsProjectItem" ) )
+      sipType = sipType_QgsProjectItem;
+    else
+      sipType = 0;
```
I think it's a bit faster ot use `qobject_cast` (as before)

doublebyte1 replied:
done

m-kuhn commented:
Thanks, good job!

doublebyte1 committed:
- updated const variables on the qgsvector header.
- Removed dataItem_t pointer from qgsdataitem bindings.
- Put back qobject_cast.

doublebyte1 committed:
- updated const variables on the qgsvector header.
- Removed dataItem_t pointer from qgsdataitem bindings.
- Put back qobject_cast.

doublebyte1 committed:
Merge remote-tracking branch 'upstream/master' into sip-annotated

doublebyte1 committed:
Merge remote-tracking branch 'upstream/master' into sip-annotated

doublebyte1 committed:
Merge branch 'sip-annotated' of https://github.com/GeoCat/QGIS into sip-annotated

3nids commented:
I updated the sipify script. You might need to rebase again, sorry!

doublebyte1 committed:
- changes to sipify script
- updated gsdataitem header:
  - added qgis.h include
  - skip sDataItemPtr
  - remove , between SIP annotations
- run sipify for qgsvector and qgsdataitem

doublebyte1 commented:
@3nids Ok. I updated the source code and pushed the changes. The build compiles now! :-)

3nids committed:
Merge branch 'master' into sip-annotated

3nids commented:
manually merged, thanks a lot

3nids commented:
I am really sorry the commits were transferred to me as author. I had troubles with the merging and did not do it properly apparently. really sorry for that.

3nids commented:
nope, it's ok, it's just the merge commits
https://github.com/qgis/QGIS/commits?author=doublebyte1

doublebyte1 commented:
@3nids great, I am glad I could help!
